mitodl/mitxpro
# Create a new auth token if the old one fails to work/refresh

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #2464 

#### What's this PR do?
- If a user's `OpenEdxApiAuth` fails to reauthenticate via refresh token, it will be deleted and replaced with a new one.
- This should fix one sentry issue that occurs when running the `retry_failed_edx_enrollments` task (but not [another caused by runs with expired enrollment dates](https://github.com/mitodl/mitxpro/issues/2470)).  

#### How should this be manually tested?
Try this in a shell:
```
import datetime
import pytz
from courseware.api import *
from courseware.api import _refresh_edx_api_auth


user = User.objects.get(id=<some_user_id>)
bad_auth = OpenEdxApiAuth.objects.get(user=user)
bad_auth.access_token="fake"
bad_auth.refresh_token="fake"
bad_auth.access_token_expires_on = datetime.datetime(2020,1,1, tzinfo=pytz.UTC)
new_auth = get_valid_edx_api_auth(user)
print(new_auth.access_token, new_auth.refresh_token) # Should not be 'fake'
new_auth_2 = get_valid_edx_api_auth(user)
print(new_auth_2.access_token, new_auth_2.refresh_token) # Should be same as above
new_auth_2.access_token_expires_on = datetime.datetime(2020,1,1, tzinfo=pytz.UTC)
new_auth_2.save()
new_auth_3 = get_valid_edx_api_auth(user)
print(new_auth_3.access_token, new_auth_3.refresh_token) # Should be different from above
new_auth_4 =  _refresh_edx_api_auth(new_auth_3)
print(new_auth_3.access_token, new_auth_3.refresh_token) # Should be different from above
```

Try registering, logging in, logging out, logging back in, etc.  There should be no problems doing so.
